### PR TITLE
Cloud tier restrictions

### PR DESCRIPTION
**Pull request by luciipak**

### CODE REVIEW

File: deployments/pulse3d/services/pulse3d/src/main.py [hunk 1]
```diff
@@ -330,7 +388,11 @@ def _format_tuple_param(
 async def soft_delete_jobs(
     request: Request,
     job_ids: List[uuid.UUID] = Query(None),
-    token=Depends(ProtectedAny(scope=["users:free", "users:admin"])),
+    token=Depends(
+        ProtectedAny(
+            scope=["pulse3d:user:free", "pulse3d:user:paid", "pulse3d:customer:free", "pulse3d:customer:paid"]
```
Comment (jperson) [inline deployments/pulse3d/services/pulse3d/src/main.py:393]: might want to create a list of scopes for pulse3d since this gets re-used a bunch, something like

```python
PULSE3D_SCOPES = ["pulse3d:user:free", ...]
```

File: deployments/apiv2/services/users/src/main.py [hunk 2]
```diff
@@ -236,7 +245,7 @@ async def logout(request: Request, token=Depends(ProtectedAny(check_scope=False)
 async def register(
     request: Request,
     details: Union[CustomerCreate, UserCreate],
-    token=Depends(ProtectedAny(scope=["users:admin"])),
+    token=Depends(ProtectedAny(scope=PULSE3D_CUSTOMER_SCOPES)),
```
Comment (jperson) [inline deployments/apiv2/services/users/src/main.py:248]: I don't think this route should be specific to pulse3d users, the customer admin account won't be tied to a specific product

File: deployments/apiv2/services/users/src/main.py [hunk 3]
```diff
@@ -369,7 +387,10 @@ async def _send_registration_email(username: str, email: EmailStr, verification_
 
 
 @app.get("/")
-async def get_all_users(request: Request, token=Depends(ProtectedAny(scope=["users:admin"]))):
+async def get_all_users(
+    request: Request,
+    token=Depends(ProtectedAny(scope=PULSE3D_CUSTOMER_SCOPES)),
```
[the remark this answers is not in the document]
Comment (luciipak) [reply deployments/apiv2/services/users/src/main.py:392]: Got it, yeah I didn't add anything yet for a super admin account for curibio.. and was thinking customers would have service scopes in case they didn't pay for all services so thought we'd just add `ProtectedAny(scope=[PULSE3D_CUSTOMER_SCOPES, <SERVICE>_CUSTOMER_SCOPES, ....]`

[the remark this answers is not in the document]
Comment (luciipak) [reply deployments/apiv2/services/users/src/main.py:392]: is a customer account logged into the dashboard page specific to pulse3d or will it be attached to all the services?

File: core/db/curibio/alembic/versions/da5a93c1bb08_update_uploadtype_and_scope_schema.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+
+def upgrade():
+    # default all existing customers and users to paid pulse3d
+    op.execute(f"update users set data='{json.dumps({'scope': ['pulse3d:user:paid']})}'")
+    op.execute(f"update customers set data='{json.dumps({'scope': ['pulse3d:customer:paid']})}'")
+
+    # change mantarray uploadtype to pulse3d
+    # set existing mantarray types to other accepted type that is not currently being used
+    op.execute("""UPDATE uploads SET type='nautilus' WHERE type='mantarray'""")
```
Comment (tannermpeterson) [inline core/db/curibio/alembic/versions/da5a93c1bb08_update_uploadtype_and_scope_schema.py:25]: Do these need to be in triple quotes?

Comment (luciipak) [reply core/db/curibio/alembic/versions/da5a93c1bb08_update_uploadtype_and_scope_schema.py:25]: Nah, just noticed all the other migrations had it too, i'll remove them

File: core/lib/auth/auth/routes.py [hunk 5]
```diff
@@ -73,8 +75,8 @@ def create_token(
         raise ValueError(f"Valid account types are 'user' and 'customer', not '{account_type}'")
     if account_type == "user":
         # make sure a user is not given admin privileges
-        if "users:admin" in scope:
-            raise ValueError("User tokens cannot have scope 'users:admin'")
+        if "customer" in scope:
```
Comment (tannermpeterson) [inline core/lib/auth/auth/routes.py:78]: The scope is a list, not a string, so you'll probably want to check each item in the list

File: core/lib/auth/auth/scopes.py [hunk 6]
```diff
@@ -0,0 +1,3 @@
+PULSE3D_CUSTOMER_SCOPES = ["pulse3d:customer:free", "pulse3d:customer:paid"]
```
Comment (tannermpeterson) [inline core/lib/auth/auth/scopes.py:1]: These should be tuples so that they're not mutable

File: core/lib/auth/setup.py [hunk 7]
```diff
@@ -2,7 +2,7 @@
 
 setup(
     name="auth",
-    version="0.0.1",
+    version="0.0.2",
```
Comment (tannermpeterson) [inline core/lib/auth/setup.py:5]: Is this version even used anywhere?

File: core/lib/jobs/jobs/jobs.py [hunk 8]
```diff
@@ -213,3 +217,30 @@ def _get_placeholders_str(num_placeholders, start=1):
...
+        - Dictionary containing boolean values for if uploads or job quotas have been reached
+    """
+
+    async with con.transaction():
+        # get service specific usage restrictions for the customer account
+        usage_query = "select usage_restrictions->$1 as usage from customers where id=$2"
+        usage_json = await con.fetchrow(usage_query, service, customer_id)
+        usage_dict = json.loads(usage_json["usage"])
+        # grab total uploads and jobs of customer for specific service
+        jobs_type = "mantarray"  # REMEMBER TO CHANGE BACK AFTER MIGRATING TO USE PULSE3D
```
Comment (tannermpeterson) [inline core/lib/jobs/jobs/jobs.py:236]: This should be a TODO

File: core/lib/auth/setup.py (hunk 7 above)
Comment (luciipak) [reply core/lib/auth/setup.py:5]: not sure tbh

File: core/lib/jobs/jobs/jobs.py (hunk 8 above)
Comment (luciipak) [reply core/lib/jobs/jobs/jobs.py:236]: I'm changing this as soon as I run the migration before I merge this pr, so I'm going to push that up

File: core/lib/jobs/jobs/jobs.py [hunk 9]
```diff
@@ -213,3 +217,30 @@ def _get_placeholders_str(num_placeholders, start=1):
...
+        query = "select count(*) from jobs_result where customer_id=$1 and type=$2 group by upload_id"
```
Comment (tannermpeterson) [inline core/lib/jobs/jobs/jobs.py:237]: nitpick, should capitalize all the SQL keywords in these queries

File: core/lib/jobs/jobs/jobs.py [hunk 10]
```diff
@@ -213,3 +217,30 @@ def _get_placeholders_str(num_placeholders, start=1):
...
+        customer_data = [row["count"] async for row in con.cursor(query, customer_id, jobs_type)]
+        total_uploads = len(customer_data)
+        total_jobs = sum(customer_data)
```
Comment (tannermpeterson) [inline core/lib/jobs/jobs/jobs.py:240]: Not sure how important this is, but I think you could calculate these two values in the sql query

File: deployments/apiv2/services/users/src/main.py [hunk 11]
```diff
@@ -95,15 +97,16 @@ async def login(request: Request, details: Union[UserLogin, CustomerLogin]):
...
         account_type = "user"
         # suspended is for deactivated accounts and verified is for new users needing to verify through email
-        select_query = (
-            "SELECT password, id, data->'scope' AS scope "
-            "FROM users WHERE deleted_at IS NULL AND name=$1 AND customer_id=$2 AND suspended='f' AND verified='t'"
+        # select for service specific usage restrictions listed under the customer account
+        select_query = "SELECT password, id, data->'scope' AS scope FROM users WHERE deleted_at IS NULL AND name=$1 AND customer_id=$2 AND suspended='f' AND verified='t'"
+        select_query_params = (
+            details.username,
+            str(details.customer_id),
```
Comment (tannermpeterson) [inline deployments/apiv2/services/users/src/main.py:104]: Were these formatting changes necessary? the assignment to `select_query` was written the original way where it wraps multiple lines so it's more readable

File: deployments/apiv2/services/users/src/models/users.py [hunk 12]
```diff
@@ -85,14 +92,19 @@ class UserProfile(BaseModel):
...
     user_id: str
-    scope: list
+    scope: List[str]
 
 
 class UserAction(BaseModel):
     action_type: str
+
+
+class LoginResponse(BaseModel):
```
Comment (tannermpeterson) [inline deployments/apiv2/services/users/src/models/users.py:108]: I think we'll need to update the desktop app with this change, and potentially some of the other changes as well

File: deployments/apiv2/services/users/src/main.py (hunk 11 above)
Comment (luciipak) [reply deployments/apiv2/services/users/src/main.py:104]: did the hooks not  do this? I don't remember if this was me or the formatting hooks, but can  switch it back yes

File: frontend/pulse3d/src/components/layouts/ControlPanel.js [hunk 13]
```diff
@@ -118,82 +103,148 @@ const adminButtons = [
...
 
   useEffect(() => {
+    // this checks if a page changes without button is clicked from a forced redirection
     const currentPage = buttons.filter(({ page }) => page === router.pathname)[0];
 
     if (currentPage) {
       const { label, options } = currentPage;
       if (label !== selected) setSelected(label);
       if (options.length > 0) setExpanded(label);
+      else if (options.length === 0) setExpanded(null);
```
Comment (tannermpeterson) [inline frontend/pulse3d/src/components/layouts/ControlPanel.js:160]: This could just be `else setExpanded(null);` unless `options.length` can be < 0 somehow

File: frontend/pulse3d/src/pages/uploads.js [hunk 14]
```diff
@@ -247,7 +249,7 @@ export default function Uploads() {
             checked: isChecked,
           };
         });
-        setJobs(newJobs);
+        setJobs([...newJobs]);
```
Comment (tannermpeterson) [inline frontend/pulse3d/src/pages/uploads.js:252]: Why is this change necessary? Just curious

File: deployments/apiv2/services/users/src/main.py (hunk 11 above)
Comment (tannermpeterson) [reply deployments/apiv2/services/users/src/main.py:104]: Oh yeah, not sure if the pre-commit hooks would change this

File: core/lib/jobs/jobs/jobs.py (hunk 10 above)
Comment (luciipak) [reply core/lib/jobs/jobs/jobs.py:240]: I'll look into it, this will likely be changed soon when we go into a rolling monthly quota

File: deployments/apiv2/services/users/src/models/users.py (hunk 12 above)
Comment (luciipak) [reply deployments/apiv2/services/users/src/models/users.py:108]: before merging this to main right?

Comment (tannermpeterson) [reply deployments/apiv2/services/users/src/models/users.py:108]: Yeah we should just make a ticket to do it

File: frontend/pulse3d/src/pages/uploads.js (hunk 14 above)
Comment (luciipak) [reply frontend/pulse3d/src/pages/uploads.js:252]: Oh no.. I think i did this when I was trying to see why the loading spinner never disappeared and why the table was rerendering so often

Comment (luciipak) [reply frontend/pulse3d/src/pages/uploads.js:252]: good catch, I'll swap it back. It doesn't matter
